pynitrokey/cli/update.py

```diff
@@ -7,45 +7,96 @@
-@click.option(
-    "-y", "--yes", is_flag=True, help="Don't ask for confirmation before flashing"
-)
-@click.option(
-    "-lfs",
-    "--local-firmware-server",
-    is_flag=True,
-    default=False,
-    hidden=True,
-    help="Development option: pull firmware from http://localhost:8000",
-)
-@click.option(
-    "--alpha",
-    is_flag=True,
-    default=False,
-    hidden=True,
-    help="Development option: use release refered to by ALPHA_VERSION",
-)
-def update(serial, yes, local_firmware_server, alpha):
+@click.option("-s", "--serial", help="Serial number of Nitrokey key to target",
+              default=None)
+def update(serial):
     """Update Nitrokey key to latest firmware version."""
 
-    update_url = 'https://update.nitrokey.com/'
-    print('Please use {} to run the firmware update'.format(update_url))
-    return
+    #update_url = 'https://update.nitrokey.com/'
+    #print('Please use {} to run the firmware update'.format(update_url))
+    #return
+
+    # Determine target key
+    try:
+        client = pynitrokey.client.find(serial)
+
+    except pynitrokey.exceptions.NoSoloFoundError:
+        print()
+        print("No Nitrokey key found!")
+        print()
+        print("If you are on Linux, are your udev rules up to date?")
+        print("For more, see https://www.nitrokey.com/documentation/installation#os:linux")
+        print()
+        sys.exit(1)
+    except pynitrokey.exceptions.NonUniqueDeviceError:
+        print()
+        print("Multiple Nitrokey keys are plugged in! Please:")
+        print("  * unplug all but one key")
+        print()
+        sys.exit(1)
+    except Exception:
+        print()
+        print("Unhandled error connecting to key.")
+        print("Please report via https://github.com/Nitrokey/pynitrokey/issues/")
+        print()
+        sys.exit(1)
+
+    # determine asset url: we want the (signed) json file
+    api_url = "https://api.github.com/repos/Nitrokey/nitrokey-fido2-firmware/releases/latest"
+    assets = [(x["name"], x["browser_download_url"])
+              for x in json.loads(requests.get(api_url).text)["assets"]]
+    download_url = None
+    for fn, url in assets:
+        if fn.endswith(".json"):
+            download_url = url
+            break
+    if download_url is None:
+        print("Failed to determine latest release")
+        return
+
+    # download asset url
+    print("Downloading latest firmware")
+    tmp_dir = tempfile.gettempdir()
+    fw_fn = os.path.join(tmp_dir, "fido2_firmware.json")
+    with open(fw_fn, "wb") as fd:
+        firmware = requests.get(download_url)
+        fd.write(firmware.content)
+
+    # Ensure we are in bootloader mode
+    if client.is_solo_bootloader():
+        print("Key already in bootloader mode, continuing...")
+    else:
+        print("Entering bootloader mode, please confirm with button on key!")
+        client.enter_bootloader_or_die()
+        time.sleep(0.5)
```

Comment:
I would add here try-catch block for time-out of bootloader switching, otherwise this will end up in nasty stacktrace log.